From: Fab <netbox253@gmail.com>

The following patch is originally from gentoo-sources repository [1] (3.4 branch).
It is adapted to apply to dan-and's linux-sunxi repository [2]
(dan-3.4.current branch) after commit b095fd3337f150c56167e1871a09e01103f5ef28
was applied. Original comment follows.

1. https://gitweb.gentoo.org/proj/linux-patches.git/tree/?h=3.4
2. https://github.com/dan-and/linux-sunxi

===

From: Anthony G. Basile <blueness@gentoo.org>

This patch adds support for a restricted user-controlled namespace on
tmpfs filesystem used to house PaX flags.  The namespace must be of the
form user.pax.* and its value cannot exceed a size of 8 bytes.

This is needed even on all Gentoo systems so that XATTR_PAX flags
are preserved for users who might build packages using portage on
a tmpfs system with a non-hardened kernel and then switch to a
hardened kernel with XATTR_PAX enabled.

The namespace is added to any user with Extended Attribute support
enabled for tmpfs.  Users who do not enable xattrs will not have
the XATTR_PAX flags preserved.

diff --git a/include/linux/xattr.h b/include/linux/xattr.h
index 2ace7a6..7331b27 100644
--- a/include/linux/xattr.h
+++ b/include/linux/xattr.h
@@ -146,4 +146,9 @@ void simple_xattr_list_add(struct simple_xattrs *xattrs,
 
 #endif  /*  __KERNEL__  */
 
+/* User namespace */
+#define XATTR_PAX_PREFIX XATTR_USER_PREFIX "pax."
+#define XATTR_PAX_FLAGS_SUFFIX "flags"
+#define XATTR_NAME_PAX_FLAGS XATTR_PAX_PREFIX XATTR_PAX_FLAGS_SUFFIX
+
 #endif	/* _LINUX_XATTR_H */
diff --git a/mm/shmem.c b/mm/shmem.c
index 03a6e1b..84a0295 100644
--- a/mm/shmem.c
+++ b/mm/shmem.c
@@ -1881,6 +1881,7 @@ static const struct xattr_handler *shmem_xattr_handlers[] = {
 static int shmem_xattr_validate(const char *name)
 {
 	struct { const char *prefix; size_t len; } arr[] = {
+		{ XATTR_USER_PREFIX, XATTR_USER_PREFIX_LEN},
 		{ XATTR_SECURITY_PREFIX, XATTR_SECURITY_PREFIX_LEN },
 		{ XATTR_TRUSTED_PREFIX, XATTR_TRUSTED_PREFIX_LEN }
 	};
@@ -1936,6 +1937,13 @@ static int shmem_setxattr(struct dentry *dentry, const char *name,
 	if (err)
 		return err;
 
+	if (!strncmp(name, XATTR_USER_PREFIX, XATTR_USER_PREFIX_LEN)) {
+		if (strcmp(name, XATTR_NAME_PAX_FLAGS))
+			return -EOPNOTSUPP;
+		if (size > 8)
+			return -EINVAL;
+	}
+
 	return simple_xattr_set(&info->xattrs, name, value, size, flags);
 }
 
